galley/pkg/runtime/projections/serviceentry/handler.go

```diff
@@ -0,0 +1,302 @@
+	endpoints        map[resource.FullName]resource.Entry
+	podsHandler      processing.Handler
+	nodesHandler     processing.Handler
+	localityStrategy locality.Strategy
+
+	// The version number for the current State of the object. Every time mcpResources or versions change,
+	// the version number also change
+	version      int64
+	mcpLock      sync.Mutex
+	mcpResources map[resource.FullName]*mcp.Resource
+
+	// pendingEvents counts the number of events awaiting publishing.
+	pendingEvents int64
+
+	// lastSnapshotTime records the last time a snapshot was published.
+	lastSnapshotTime time.Time
+}
+
+// NewHandler creates a new Handler instance.
+func NewHandler(domainSuffix string, listener processing.Listener) *Handler {
+	pods, podsHandler := pod.NewCache()
+	nodes, nodesHandler := node.NewCache()
+
+	return &Handler{
+		domainSuffix:     domainSuffix,
+		listener:         listener,
+		services:         make(map[resource.FullName]resource.Entry),
+		endpoints:        make(map[resource.FullName]resource.Entry),
+		localityStrategy: locality.NewStrategy(pods, nodes),
+		podsHandler:      podsHandler,
+		nodesHandler:     nodesHandler,
+	}
+}
+
+// Handle incoming events and generate synthetic ServiceEntry projections.
+func (p *Handler) Handle(event resource.Event) {
+	if event.Kind == resource.FullSync {
+		p.distribute = true
+		p.notifyChanged()
+		return
+	}
+
+	switch event.Entry.ID.Collection {
+	case metadata.Endpoints.Collection:
+		// Update the projections
+		p.handleEndpointsEvent(event)
+	case metadata.Service.Collection:
+		// Update the projections
+		p.handleServiceEvent(event)
+	case metadata.Node.Collection:
+		// Just add the node to the cache.
+		p.nodesHandler.Handle(event)
+	case metadata.Pod.Collection:
+		// Just add the pod to the cache.
+		p.podsHandler.Handle(event)
+	default:
+		scope.Warnf("received event with unexpected collection: %v", event.Entry.ID.Collection)
+	}
+}
+
+// Builds the snapshot of the current resources.
+func (p *Handler) BuildSnapshot() snapshot.Snapshot {
+	p.mcpLock.Lock()
+	defer p.mcpLock.Unlock()
+
+	now := time.Now()
+	monitoring.RecordProcessorSnapshotPublished(p.pendingEvents, now.Sub(p.lastSnapshotTime))
+	p.lastSnapshotTime = now
+
+	b := snapshot.NewInMemoryBuilder()
+
+	entries := make([]*mcp.Resource, 0, len(p.mcpResources))
+	for _, r := range p.mcpResources {
+		entries = append(entries, r)
+	}
+	version := fmt.Sprintf("%d", p.version)
+	b.Set(collection.String(), version, entries)
+
+	p.pendingEvents = 0
+
+	return b.Build()
+}
+
+func (p *Handler) handleServiceEvent(event resource.Event) {
+	entry := event.Entry
+	fullName := entry.ID.FullName
+
+	switch event.Kind {
+	case resource.Added, resource.Updated:
+
+		// Check to see if the service's version has changed.
+		curEntry, ok := p.services[fullName]
+		if ok && curEntry.ID.Version == entry.ID.Version {
+			scope.Debugf("received service for the current, known version: %v", event)
```

Comment:
Nit: I'm wondering if this debug line is going to cause a bit of noise since add/update is not often. Would it make sense to have the debug when we add/update?